File: src/audio/playback/player/GeneralPlayer.java
package audio.playback.player;

import javafx.scene.media.MediaPlayer.Status;
import javafx.util.Duration;

import audio.Item;
import audio.Player;
import audio.playback.PLAYBACK;
import audio.playback.PlayTimeHandler;
import audio.playback.PlaybackState;
import audio.playback.RealTimeProperty;
import audio.playlist.Playlist;
import audio.playlist.PlaylistItem;
import audio.playlist.PlaylistManager;
import audio.tagging.Metadata;
import util.animation.Anim;
import util.async.Async;
import util.file.AudioFileFormat;

import static audio.playback.PLAYBACK.*;
import static java.lang.Math.pow;
import static javafx.scene.media.MediaPlayer.Status.PLAYING;
import static javafx.scene.media.MediaPlayer.Status.STOPPED;
import static javafx.util.Duration.ZERO;
import static javafx.util.Duration.millis;
import static util.async.Async.runFX;
import static util.dev.Util.log;
import static util.file.AudioFileFormat.*;
import static util.reactive.Util.installSingletonListener;

/**
 * Audio player which abstracts away from the implementation. It 'uses' actual players and provides
 * some of the higher level features they may lack.
 *
 * @author Martin Polakovic
 */
public class GeneralPlayer {

    private final JavaSoundPlayer fp = new JavaSoundPlayer();
    private final JavaFxPlayer dp = new JavaFxPlayer();
//    private final GPlayer gp = new GPlayer();
    private Play p;

    private Play getPlayer(Item i) {
//        return gp;
        AudioFileFormat f = i.getFormat();
        if (!f.isSupported(Use.PLAYBACK)) return null;
        if (f==mp3 || f==wav || f==mp4 || f==m4a) return dp;
        else return fp;
    }

    Item i;
    public final RealTimeProperty realTime = new RealTimeProperty(state.duration, state.currentTime);

    @SuppressWarnings("deprecation")
    synchronized public void play(PlaylistItem item) {
        // Do not recreate player if same song plays again
        // 1) improves performance
        // 2) avoids firing some playback events
        if (p!=null && item.same(i)) {
            seek(ZERO);
            return;
        }

        i = item;
        if (p!=null) p.dispose();
        p = getPlayer(item);

        try {
            if (p==null) throw new NoPlayerException();
            p.createPlayback(item, state,
                () -> {
                    realTime.real_seek = state.realTime.get();
                    realTime.curr_sek = ZERO;
                    p.play();

                    realTime.synchroRealTime_onPlayed();
                    // throw item change event
                    Player.playingItem.itemChanged(item);
                    suspension_flag = false;
                    // fire other events (may rely on the above)
                    PLAYBACK.onPlaybackStart.run();
                    if (post_activating_1st || !post_activating)
                        // bug fix, not updated playlist items can get here, but should not!
                        if (item.getTimeMs()>0)
                            PLAYBACK.onPlaybackAt.forEach(t -> t.restart(item.getTime()));
                    post_activating = false;
                    post_activating_1st = false;
                },
                () -> runFX(() -> {
                    log(GeneralPlayer.class).info("Player {} can not play item {}", p,item);
                    item.playbackerror = true;
                    PlaylistManager.use(Playlist::playNextItem); // handle within playlist
                }));
        } catch(NoPlayerException e) {
            log(GeneralPlayer.class).info("Player {} can not play item {}", p,item);
            item.playbackerror = true;
            PlaylistManager.use(Playlist::playNextItem); // handle within playlist
        }
    }

    public void resume() {
        if (p==null) return;
        p.resume();
        PLAYBACK.onPlaybackAt.forEach(PlayTimeHandler::unpause);
    }

    public void pause() {
        if (p==null) return;
        p.pause();
        PLAYBACK.onPlaybackAt.forEach(PlayTimeHandler::pause);
    }

    public void pause_resume() {
        if (p==null) return;
        if (state.status.get()==PLAYING) pause();
        else resume();
    }

    public void stop() {
        if (p==null) return;
        p.stop();

        Async.runFX(() -> {
            Player.playingItem.itemChanged(Metadata.EMPTY);
            realTime.synchroRealTime_onStopped();
            PLAYBACK.onPlaybackAt.forEach(PlayTimeHandler::stop);
            PlaylistManager.playlists.forEach(p -> p.updatePlayingItem(-1));
            PlaylistManager.active = null;
        });
    }

    public void seek(Duration duration) {
        if (p==null) return;

        Status s = state.status.get();

        if (s == STOPPED) {
            pause();
        }

        final double currentVolume = state.volume.get();
        if (seekDone) lastValidVolume = currentVolume;
        else { if (volumeAnim!=null) volumeAnim.pause(); }
        seekDone = false;
        new Anim(millis(150), x -> state.volume.set(currentVolume*pow(1-x,2)))
                .then(() -> {
                        doSeek(duration);
                        volumeAnim = new Anim(millis(150), x -> state.volume.set(lastValidVolume*pow(x,2)))
                                .then(() -> seekDone=true);
                        volumeAnim.playOpen();
                })
                .playOpen();
    }

    private void doSeek(Duration duration) {
        realTime.synchroRealTime_onPreSeeked();
        installSingletonListener(state.currentTime, v -> v!=null, v -> PLAYBACK.onSeekDone.run());
        p.seek(duration);
        realTime.synchroRealTime_onPostSeeked(duration);
    }

    private boolean seekDone = true;
    private double lastValidVolume = -1;
    private Anim volumeAnim;

    /**
     * Closes player. Releases resources. Prevents double playback.
     */
    public void dispose() {
        if (p==null) return;
        p.dispose();
        p=null;
    }

    public interface Play {
        void play();
        void pause();
        void resume();
        void seek(Duration duration);
        void stop();
        void createPlayback(Item item, PlaybackState state, Runnable onOK, Runnable onFail);
        /**
         * Stops playback if any and disposes of the player resources.
         */
        void dispose();
    }
    private static class NoPlayerException extends Exception {}
}

File: src/player/main/sp/it/pl/audio/playback/GeneralPlayer.kt
package sp.it.pl.audio.playback

import javafx.beans.property.ReadOnlyObjectWrapper
import javafx.scene.media.MediaPlayer.Status.PLAYING
import javafx.scene.media.MediaPlayer.Status.STOPPED
import javafx.util.Duration
import javafx.util.Duration.ZERO
import kotlin.math.pow
import mu.KLogging
import sp.it.pl.audio.PlayerState
import sp.it.pl.audio.Song
import sp.it.pl.audio.playlist.PlaylistManager
import sp.it.pl.audio.playlist.PlaylistSong
import sp.it.pl.audio.tagging.Metadata
import sp.it.pl.main.APP
import sp.it.pl.main.AppTexts.textNoVal
import sp.it.util.access.readOnly
import sp.it.util.animation.Anim.Companion.anim
import sp.it.util.async.runFX
import sp.it.util.async.runVT
import sp.it.util.functional.ifNotNull
import sp.it.util.functional.ifNull
import sp.it.util.reactive.attach1IfNonNull
import sp.it.util.units.millis

/** Audio player which abstracts away from the implementation. */
class GeneralPlayer(state: PlayerState) {

   private val state = state
   private var p: Play? = null
   private val _pInfo = ReadOnlyObjectWrapper("<none>")
   val pInfo = _pInfo.readOnlyProperty.readOnly()
   private var i: Song? = null
   private var seekVolumeAnim = anim(150.millis) { state.playback.volumeFadeMultiplier.value = it.pow(2) }

   fun play(song: PlaylistSong) {
      APP.audio.isSuspendedBecauseStartedPaused = false

      // Do not recreate player if same song plays again
      // 1) improves performance
      // 2) avoids firing some playback events
      if (p!=null && song.same(i)) {
         seek(ZERO)
         return
      }

      i = song
      p?.disposePlayback()
      val player = p ?: computePlayer().apply { p = this }
      val onUnableToPlay = { _: PlaylistSong -> runFX { PlaylistManager.use { it.playNextItem() } } }

      _pInfo.value = when (p) {
         null -> textNoVal
         is VlcPlayer -> "VlcPlayer"
         is JavaFxPlayer -> "JavaFX"
      }

      if (player==null) {
         logger.info { "Player=null can not play song=$song{}" }
         onUnableToPlay(song)
      } else {
         runVT {
            if (song.isCorrupt()) {
               onUnableToPlay(song)
            } else {
               runFX {
                  player.createPlayback(
                     song,
                     state.playback,
                     onOK = {
                        player.play()
                        state.playback.realTimeImpl.syncRealTimeOnPlay()
                        // throw song change event
                        APP.audio.playingSong.songChanged(song)
                        APP.audio.isSuspended = false
                        // fire other events (may rely on the above)
                        APP.audio.onPlaybackStart()
                        if (APP.audio.postActivating1st || !APP.audio.postActivating)
                        song.time?.takeIf { it.toMillis()>0.0 }.ifNotNull { time -> APP.audio.onPlaybackAt.forEach { it.restart(time) } } // TODO: this should work for corrupted/ not updated or 00:00 songs
                        APP.audio.postActivating = false
                        APP.audio.postActivating1st = false
                     },
                     onFail = { unableToPlayAny ->
                        logger.info { "Player=$p can not play song=$song" }
                        if (unableToPlayAny) {
                           stop()
                           // TODO: notify user
                        } else {
                           onUnableToPlay(song)
                        }
                     }
                  )
               }
            }
         }
      }
   }

   @Suppress("RedundantNullableReturnType")
   private fun computePlayer(): Play? = VlcPlayer()

   fun resume() {
      p.ifNotNull {
         it.resume()
         APP.audio.onPlaybackAt.forEach { it.unpause() }
      }.ifNull {
         if (APP.audio.isSuspendedBecauseStartedPaused)
            PlaylistManager.use {
               it.playing.ifNotNull { play(it) }
            }
      }
   }

   fun pause() {
      p.ifNotNull {
         it.pause()
         APP.audio.onPlaybackAt.forEach { it.pause() }
      }
   }

   fun pauseResume() {
      if (state.playback.status.value==PLAYING) pause()
      else resume()
   }

   fun stop() {
      p.ifNotNull {
         it.stop()
         runFX {
            APP.audio.playingSong.songChanged(Metadata.EMPTY)
            APP.audio.onPlaybackAt.forEach { it.stop() }
            PlaylistManager.playlists.forEach { it.updatePlayingItem(-1) }
            PlaylistManager.active = null
         }
      }
   }

   fun seek(to: Duration) {
      p.ifNotNull {
         if (state.playback.status.value==STOPPED) pause()

         seekVolumeAnim.stop()
         state.playback.volumeFadeMultiplier.value = 0.0
         state.playback.currentTime.value = to // allow next doSeek() target correct value even if this has not finished
         state.playback.currentTime.attach1IfNonNull { APP.audio.onSeekDone() }
         p?.seek(to)
         seekVolumeAnim.jumpTo(ZERO)
         seekVolumeAnim.playFromDir(true)
      }.ifNull {
         if (APP.audio.isSuspendedBecauseStartedPaused)
            PlaylistManager.use {
               it.playing.ifNotNull { play(it) }
            }
      }
   }

   fun disposePlayback() {
      p?.disposePlayback()
      p = null
   }

   fun dispose() {
      p?.dispose()
      p = null
   }

   companion object: KLogging()

   sealed interface Play {

      val canDoSongRepeat: Boolean

      fun play()

      fun pause()

      fun resume()

      fun seek(duration: Duration)

      fun stop()

      fun createPlayback(song: Song, state: PlaybackState, onOK: () -> Unit, onFail: (Boolean) -> Unit)

      /** Stops playback if any and disposes of the player resources. */
      fun disposePlayback()   // TODO: JavaFXPLayer has async implementation, but lcPlayer has blocking implementation, fix and document

      fun dispose()
   }

}